File: crates/ferrogate-cli/src/config/loader.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-06-11
// description: Token4AI Cloud, FerroGate AI Gateway, Rust API Gateway, agent-native AI traffic infrastructure.

use anyhow::{Context, Result as AnyResult};
use ferrogate_config::{
    is_caddyfile_path, load_caddyfile, parse_caddyfile, GatewayConfig, GatewayTlsAcmeConfig,
};
use std::path::{Path, PathBuf};
use tracing::warn;

use super::{
    AdminConfig, Config, HeaderMutation, Model, Provider, RouteRule, TelemetryConfig, TlsConfig,
    Upstream,
};

impl Config {
    pub(crate) fn load(path: &PathBuf) -> AnyResult<Self> {
        if !path.exists() {
            warn!(
                config = %path.display(),
                "configuration file not found; using built-in defaults"
            );
            return Ok(Self::default());
        }

        if is_caddyfile_path(path) {
            let mut config = Self::from_gateway_config(load_caddyfile(path)?);
            config.resolve_paths_relative_to(path.parent());
            config.materialize_skill_package_resources();
            config
                .validate()
                .with_context(|| format!("failed to validate config file {}", path.display()))?;
            return Ok(config);
        }

        let raw = std::fs::read_to_string(path)
            .with_context(|| format!("failed to read config file {}", path.display()))?;
        let mut config: Self = parse_structured_config(path, &raw)?;
        config.resolve_paths_relative_to(path.parent());
        config.materialize_skill_package_resources();
        config
            .validate()
            .with_context(|| format!("failed to validate config file {}", path.display()))?;
        Ok(config)
    }

    pub(crate) fn from_toml_str(raw: &str) -> AnyResult<Self> {
        let mut config: Self = toml::from_str(raw).context("failed to parse TOML config")?;
        config.materialize_skill_package_resources();
        config.validate()?;
        Ok(config)
    }

    pub(crate) fn from_yaml_str(raw: &str) -> AnyResult<Self> {
        let mut config: Self = serde_yaml::from_str(raw).context("failed to parse YAML config")?;
        config.materialize_skill_package_resources();
        config.validate()?;
        Ok(config)
    }

    pub(crate) fn from_caddyfile_str(raw: &str, file: &str) -> AnyResult<Self> {
        let mut config = Self::from_gateway_config(parse_caddyfile(raw, file)?);
        config.materialize_skill_package_resources();
        config.validate()?;
        Ok(config)
    }

    /// Map the Caddyfile intermediate model into the runtime config shape.
    fn from_gateway_config(config: GatewayConfig) -> Self {
        Self {
            listen: config.listen,
            admin: AdminConfig {
                listen: config.admin,
            },
            auth_service: crate::config::AuthServiceConfig::default(),
            tls: match (config.tls, config.tls_acme) {
                (Some(tls), None) => TlsConfig {
                    enabled: true,
                    cert_path: Some(tls.cert_path),
                    key_path: Some(tls.key_path),
                    http2: false,
                    acme: crate::config::TlsAcmeConfig::default(),
                },
                (None, Some(acme)) => Self::tls_from_gateway_acme(acme),
                (Some(tls), Some(_)) => TlsConfig {
                    enabled: true,
                    cert_path: Some(tls.cert_path),
                    key_path: Some(tls.key_path),
                    http2: false,
                    acme: crate::config::TlsAcmeConfig::default(),
                },
                (None, None) => TlsConfig::default(),
            },
            providers: config
                .providers
                .into_iter()
                .map(|provider| Provider {
                    name: provider.name,
                    kind: provider.kind,
                    base_url: provider.base_url,
                    api_key_env: provider.api_key_env,
                    openrouter_http_referer: provider.openrouter_http_referer,
                    openrouter_x_title: provider.openrouter_x_title,
                    enabled: true,
                })
                .collect(),
            models: config
                .models
                .into_iter()
                .map(|model| Model {
                    name: model.name,
                    provider: model.provider,
                    provider_model: model.provider_model,
                    routing_strategy: ferrogate_providers::RoutingStrategy::Priority,
                    fallbacks: Vec::new(),
                    visible_organization_ids: Vec::new(),
                    visible_project_ids: Vec::new(),
                    capabilities: model.capabilities,
                    context_window: model.context_window,
                    input_price_per_1m: model
                        .input_price_per_1m
                        .as_deref()
                        .and_then(|value| value.parse().ok()),
                    output_price_per_1m: model
                        .output_price_per_1m
                        .as_deref()
                        .and_then(|value| value.parse().ok()),
                    enabled: true,
                    cache_enabled: None,
                })
                .collect(),
            guardrails: Vec::new(),
            api_keys: config
                .api_keys
                .into_iter()
                .map(|key| crate::config::ApiKey {
                    id: key.id,
                    name: key.name,
                    key_env: key.key_env,
                    key: key.key,
                    key_hash: key.key_hash,
                    enabled: true,
                    scopes: key.scopes,
                    allowed_models: key.allowed_models,
                    denied_models: key.denied_models,
                    allowed_providers: key.allowed_providers,
                    denied_providers: key.denied_providers,
                    organization_id: None,
                    team_id: None,
                    project_id: None,
                    user_id: None,
                    monthly_token_budget: key.monthly_token_budget,
                    request_limit_per_minute: key.request_limit_per_minute,
                    expires_at_unix: None,
                    log_bodies: None,
                    cache_enabled: None,
                })
                .collect(),
            policies: Vec::new(),
            gateway_configs: Vec::new(),
            agent_workflows: Vec::new(),
            skill_packages: Vec::new(),
            prompt_templates: Vec::new(),
            plugins: Vec::new(),
            extensions: Vec::new(),
            mcp_servers: Vec::new(),
            agent_upstreams: Vec::new(),
            telemetry: TelemetryConfig::default(),
            observability: crate::config::ObservabilityConfig::default(),
            analytics: crate::config::AnalyticsConfig::default(),
            metering: crate::config::MeteringConfig::default(),
            cache: crate::config::CacheConfig::default(),
            storage: crate::config::StorageConfig::default(),
            reliability: crate::config::ReliabilityConfig::default(),
            agent_runtime: crate::config::AgentRuntimeConfig::default(),
            cluster: crate::config::ClusterConfig::default(),
            upstreams: config
                .upstreams
                .into_iter()
                .map(|upstream| Upstream {
                    name: upstream.name,
                    url: Some(upstream.url),
                    urls: upstream.urls,
                    enabled: true,
                })
                .collect(),
            routes: config
                .routes
                .into_iter()
                .filter_map(|route| {
                    route.upstream.map(|upstream| RouteRule {
                        name: route.name,
                        upstream,
                        hosts: route.hosts,
                        path_prefixes: route.path_prefixes,
                        match_headers: Vec::new(),
                        strip_prefix: route.strip_prefix,
                        add_prefix: None,
                        request_headers: route
                            .request_headers
                            .into_iter()
                            .map(|header| HeaderMutation {
                                name: header.name,
                                value: header.value,
                            })
                            .collect(),
                        response_headers: route
                            .response_headers
                            .into_iter()
                            .map(|header| HeaderMutation {
                                name: header.name,
                                value: header.value,
                            })
                            .collect(),
                        enabled: true,
                    })
                })
                .collect(),
        }
    }

    fn tls_from_gateway_acme(acme: GatewayTlsAcmeConfig) -> TlsConfig {
        let mut tls_acme = crate::config::TlsAcmeConfig {
            enabled: true,
            domains: acme.domains,
            email: acme.email,
            dns_provider: acme.dns_provider,
            dns_config: acme.dns_config,
            dns_hook_set: acme.dns_hook_set,
            dns_hook_cleanup: acme.dns_hook_cleanup,
            terms_agreed: true,
            ..crate::config::TlsAcmeConfig::default()
        };
        if let Some(directory_url) = acme.directory_url {
            tls_acme.directory_url = directory_url;
        }
        if let Some(challenge) = acme.challenge {
            tls_acme.challenge = challenge;
        }
        if let Some(http_challenge_listen) = acme.http_challenge_listen {
            tls_acme.http_challenge_listen = http_challenge_listen;
        }
        if let Some(storage_dir) = acme.storage_dir {
            tls_acme.storage_dir = storage_dir;
        }
        if let Some(renewal_window_secs) = acme.renewal_window_secs {
            tls_acme.renewal_window_secs = renewal_window_secs;
        }
        if let Some(renewal_check_interval_secs) = acme.renewal_check_interval_secs {
            tls_acme.renewal_check_interval_secs = renewal_check_interval_secs;
        }
        if let Some(renewal_retry_interval_secs) = acme.renewal_retry_interval_secs {
            tls_acme.renewal_retry_interval_secs = renewal_retry_interval_secs;
        }
        if let Some(auto_graceful_reload) = acme.auto_graceful_reload {
            tls_acme.auto_graceful_reload = auto_graceful_reload;
        }
        TlsConfig {
            enabled: true,
            cert_path: None,
            key_path: None,
            http2: false,
            acme: tls_acme,
        }
    }

    fn resolve_paths_relative_to(&mut self, base_dir: Option<&Path>) {
        let Some(base_dir) = base_dir else {
            return;
        };
        if let Some(cert_path) = &mut self.tls.cert_path {
            *cert_path = resolve_relative_path(base_dir, cert_path);
        }
        if let Some(key_path) = &mut self.tls.key_path {
            *key_path = resolve_relative_path(base_dir, key_path);
        }
        if self.tls.acme.enabled {
            self.tls.acme.storage_dir = resolve_relative_path(base_dir, &self.tls.acme.storage_dir);
            if let Some(dns_hook_set) = &mut self.tls.acme.dns_hook_set {
                *dns_hook_set = resolve_relative_path(base_dir, dns_hook_set);
            }
            if let Some(dns_hook_cleanup) = &mut self.tls.acme.dns_hook_cleanup {
                *dns_hook_cleanup = resolve_relative_path(base_dir, dns_hook_cleanup);
            }
        }
    }
}

fn parse_structured_config(path: &Path, raw: &str) -> AnyResult<Config> {
    match path
        .extension()
        .and_then(|extension| extension.to_str())
        .map(|extension| extension.to_ascii_lowercase())
        .as_deref()
    {
        Some("yaml" | "yml") => serde_yaml::from_str(raw).context("failed to parse YAML config"),
        _ => toml::from_str(raw).context("failed to parse TOML config"),
    }
}

fn resolve_relative_path(base_dir: &Path, raw: &str) -> String {
    let path = Path::new(raw);
    if path.is_absolute() {
        raw.to_string()
    } else {
        base_dir
            .join(path)
            .components()
            .collect::<PathBuf>()
            .to_string_lossy()
            .into_owned()
    }
}

File: packages/config/src/loader.ts
/**
 * Port of `ferrogate-config`'s `loader.rs` + `config/loader.rs` (inventory §5.5):
 * config-load precedence, the `[control_api]`/`[admin_api]` alias migration, and
 * the Caddyfile intermediate -> `Config` bridge.
 *
 * PORT-TODO(inventory §5.8) — PLATFORM LIMIT, NOT CLOSED (plus one scope note).
 *
 * PLATFORM LIMIT: workerd has NO FILESYSTEM. `Config::from_file` /
 * `from_toml_file` / `from_yaml_file` and `resolve_paths_relative_to` (which
 * rebases cert/key/ACME paths against the config file's directory) have no
 * expressible form — there is no path, no cwd, and no `std::fs`. Config on this
 * platform is sourced from wrangler vars, KV, D1 or a bundled asset, so the
 * entry point is `loadConfigFromObject(rawObject)` over an ALREADY-DECODED
 * object. `resolve_paths_relative_to` is additionally moot: it only ever
 * rebased manual-TLS/ACME paths, which are N/A behind Cloudflare's TLS
 * termination (see `schema/sections.ts`).
 *
 * SCOPE NOTE, NOT A PLATFORM LIMIT: `from_toml_str`/`from_yaml_str` decode a
 * STRING, which a Worker can perfectly well do — they are absent only because
 * this package carries no TOML/YAML parser dependency. Whoever adds one should
 * decode to an object and hand it to `loadConfigFromObject`; do NOT re-derive
 * the alias migration or validation, which are the parts that carry semantics.
 * The Caddyfile bridge does not need a dependency and IS ported
 * (`fromCaddyfileStr`). Pinned by `platform-limits.test.ts` > "loader".
 */
import { parseCaddyfile } from "./caddyfile/parser.js";
import type { GatewayConfig } from "./caddyfile/types.js";
import type { EnvSource } from "./secrets.js";
import { parseConfig, type Config } from "./schema/config.js";
import { validateConfig, type ValidateOptions } from "./validate.js";

/** Whether a path's filename is exactly `Caddyfile` (case-insensitive). */
export function isCaddyfilePath(path: string): boolean {
  const name = path.split(/[\\/]/).pop() ?? path;
  return name.toLowerCase() === "caddyfile";
}

/** The outcome of resolving the control-plane API alias. */
export interface AliasMigration {
  /** The raw config object with `control_api`/`admin_api` resolved into `admin_api`. */
  config: Record<string, unknown>;
  /** Migration notices (the deprecated-alias warning), for the caller to surface. */
  warnings: string[];
}

/**
 * `Config::migrate_control_plane_aliases` (issue #359): resolve the effective
 * control-plane API config from the canonical `[control_api]` section or the
 * deprecated `[admin_api]` alias. Both present is a hard error; alias-only
 * warns; neither leaves the defaults. Runs BEFORE schema validation (a Worker
 * has no serde `skip_deserializing`), so it operates on the raw object.
 */
export function migrateControlPlaneAliases(raw: Record<string, unknown>): AliasMigration {
  const config = { ...raw };
  const controlApi = config.control_api;
  const adminAlias = config.admin_api;
  const warnings: string[] = [];

  const hasControl = controlApi !== undefined && controlApi !== null;
  const hasAlias = adminAlias !== undefined && adminAlias !== null;

  if (hasControl && hasAlias) {
    throw new Error(
      "conflicting control-plane API configuration: both the canonical [control_api] section and " +
        "the deprecated [admin_api] alias are set. They map to the same FerroGate Control Plane " +
        "API service, so keep only one -- prefer [control_api] and remove [admin_api].",
    );
  }
  if (hasControl) {
    config.admin_api = controlApi;
  } else if (hasAlias) {
    warnings.push(
      "the [admin_api] config section is a deprecated alias; rename it to [control_api] (FerroGate " +
        "Control Plane API). The [admin_api] alias still works during the migration window and maps " +
        "1:1 onto [control_api].",
    );
    // admin_api already holds the alias value (the effective section).
  }
  delete config.control_api;
  return { config, warnings };
}

/**
 * Load a config from an already-parsed object (TOML/YAML/JSON decoded by the
 * caller): migrate the control-plane alias, validate the structural shape via
 * the Zod schema, then run the cross-field `validate()` gate.
 */
export function loadConfigFromObject(
  raw: Record<string, unknown>,
  options: ValidateOptions = {},
): { config: Config; warnings: string[] } {
  const migrated = migrateControlPlaneAliases(raw);
  const config = parseConfig(migrated.config);
  validateConfig(config, options);
  return { config, warnings: migrated.warnings };
}

/**
 * `Config::from_gateway_config`: map the Caddyfile intermediate model into a
 * raw runtime config object (every unspecified section left to its schema
 * default). Returns a plain object ready for {@link loadConfigFromObject}.
 */
export function fromGatewayConfig(gateway: GatewayConfig): Record<string, unknown> {
  let tls: Record<string, unknown> | undefined;
  if (gateway.tls !== null) {
    tls = { enabled: true, cert_path: gateway.tls.cert_path, key_path: gateway.tls.key_path, http2: false };
  } else if (gateway.tls_acme !== null) {
    const acme = gateway.tls_acme;
    const acmeOut: Record<string, unknown> = {
      enabled: true,
      domains: acme.domains,
      email: acme.email,
      dns_provider: acme.dns_provider,
      dns_config: acme.dns_config,
      dns_hook_set: acme.dns_hook_set,
      dns_hook_cleanup: acme.dns_hook_cleanup,
      terms_agreed: true,
    };
    if (acme.directory_url !== null) acmeOut.directory_url = acme.directory_url;
    if (acme.challenge !== null) acmeOut.challenge = acme.challenge;
    if (acme.http_challenge_listen !== null) acmeOut.http_challenge_listen = acme.http_challenge_listen;
    if (acme.storage_dir !== null) acmeOut.storage_dir = acme.storage_dir;
    if (acme.renewal_window_secs !== null) acmeOut.renewal_window_secs = acme.renewal_window_secs;
    if (acme.renewal_check_interval_secs !== null)
      acmeOut.renewal_check_interval_secs = acme.renewal_check_interval_secs;
    if (acme.renewal_retry_interval_secs !== null)
      acmeOut.renewal_retry_interval_secs = acme.renewal_retry_interval_secs;
    if (acme.auto_graceful_reload !== null) acmeOut.auto_graceful_reload = acme.auto_graceful_reload;
    tls = { enabled: true, cert_path: null, key_path: null, http2: false, acme: acmeOut };
  }

  const parsePrice = (value: string | null): number | null => {
    if (value === null) return null;
    const parsed = Number.parseFloat(value);
    return Number.isNaN(parsed) ? null : parsed;
  };

  return {
    listen: gateway.listen,
    admin: { listen: gateway.admin, cors_allowed_origin: null },
    ...(tls !== undefined ? { tls } : {}),
    providers: gateway.providers.map((provider) => ({
      name: provider.name,
      kind: provider.kind,
      base_url: provider.base_url,
      api_key_env: provider.api_key_env,
      openrouter_http_referer: provider.openrouter_http_referer,
      openrouter_x_title: provider.openrouter_x_title,
      enabled: true,
    })),
    models: gateway.models.map((model) => ({
      name: model.name,
      provider: model.provider,
      provider_model: model.provider_model,
      routing_strategy: "priority",
      capabilities: model.capabilities,
      context_window: model.context_window,
      input_price_per_1m: parsePrice(model.input_price_per_1m),
      output_price_per_1m: parsePrice(model.output_price_per_1m),
      enabled: true,
    })),
    api_keys: gateway.api_keys.map((key) => ({
      id: key.id,
      name: key.name,
      key_env: key.key_env,
      key: key.key,
      key_hash: key.key_hash,
      enabled: true,
      scopes: key.scopes,
      allowed_models: key.allowed_models,
      denied_models: key.denied_models,
      allowed_providers: key.allowed_providers,
      denied_providers: key.denied_providers,
      // #540: carry the DECLARED tenancy across; a Caddyfile that says neither
      // lands on null/null and is refused by validate().
      organization_id: key.organization_id,
      platform_operator: key.platform_operator,
    })),
    upstreams: gateway.upstreams.map((upstream) => ({
      name: upstream.name,
      url: upstream.url,
      urls: upstream.urls,
      enabled: true,
    })),
    routes: gateway.routes
      .filter((route) => route.upstream !== null)
      .map((route) => ({
        name: route.name,
        upstream: route.upstream,
        hosts: route.hosts,
        path_prefixes: route.path_prefixes,
        strip_prefix: route.strip_prefix,
        request_headers: route.request_headers,
        response_headers: route.response_headers,
        enabled: true,
      })),
    // #542: a Caddyfile CAN say "this gateway is open" via `auth off`.
    auth: { disabled: gateway.auth_disabled },
  };
}

/** `Config::from_caddyfile_str`: parse a Caddyfile string then load it. */
export function fromCaddyfileStr(
  raw: string,
  file: string,
  options: ValidateOptions & { env?: EnvSource } = {},
): { config: Config; warnings: string[] } {
  const gateway = parseCaddyfile(raw, file, options.env);
  return loadConfigFromObject(fromGatewayConfig(gateway), options);
}
